api/app/invitationletter/generator.py

```diff
@@ -1,9 +1,107 @@
+from mailmerge import MailMerge
+from app import LOGGER
+from config import GCP_CREDENTIALS_DICT
+from config import GCP_BUCKET_NAME
+import json
+from google.cloud import storage
+import os
+from app.utils import emailer
+from app.utils import pdfconvertor
+from app.events.models import Event
+from app import db
+from app.utils import errors
+
+os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = json.dumps(GCP_CREDENTIALS_DICT)
+
+
+INVITATION_EMAIL_BODY = """
+Dear {user_title} {first_name} {last_name},
+
+Your official invitation letter is attached!
+
+If you have any queries, please forward them to info@deeplearningindaba.com  
+
+Kind Regards,
+The Deep Learning Indaba Team
+"""
+
+
+def download_blob(bucket_name, source_blob_name, destination_file_name):
+    """Downloads a blob from the bucket."""
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(bucket_name)
+    blob = bucket.blob(source_blob_name)
+
+    blob.download_to_filename(destination_file_name)
+
+    print('Blob {} downloaded to {}.'.format(
+        source_blob_name,
+        destination_file_name))
+
+    return source_blob_name
 
 
 def generate(template_path, event_id, work_address, addressed_to, residential_address, passport_name,
              passport_no, passport_issued_by, invitation_letter_sent_at, to_date, from_date, country_of_residence,
-             nationality, date_of_birth, email):
+             nationality, date_of_birth, email, user_title, firstname, lastname, bringing_poster, expiry_date):
+
+    template = 'app/invitationletter/template/template.docx'
+    template_merged = 'app/invitationletter/letter/template.docx'
+
+    download_blob(bucket_name=GCP_BUCKET_NAME, source_blob_name=template_path,
+                  destination_file_name=template)
+
+    if os.path.exists(template):
+        document = MailMerge(template)
+        LOGGER.debug("merge-fields.... {} .".format(document.get_merge_fields()))
+
+        document.merge(
+            TITLE=user_title,
+            FIRSTNAME=firstname,
+            SURNAME=lastname,
+            WORK_ADDRESS=work_address,
+            ADDRESSED_TO=addressed_to,
+            RESIDENTIAL_ADDRESS=residential_address,
+            PASSPORT_NAME=passport_name,
```

Comment:
I think having ```Firstname```, ```Surname``` and ```Passport_Name``` is redundant.  I think we should just use ```Passport_Name```.

Reply:
The templates are currently using firstname, lastname. We don't even use passport name, but we should be

Reply:
Can't we just replace firstname and lastname, with Passport_name? We just need their full name as  shown on their passport. 

Even on the templates we can just change it to use to Passport_name. I can't think of a use case for having the names split and having seperate fields for firstname,lastname and passport name.

Reply:
Agreed. Can we merge for now though? Testing the download from gcp is proving to be nightmare, so would like to test on staging if we can.

Reply:
Cool.